### PR TITLE
Sketcher: BSpline DSH: implement OVP/widget

### PR DESCRIPTION
**Pull request by PaddleStroke**
Fixes https://github.com/FreeCAD/FreeCAD/issues/12016
Fixes https://github.com/FreeCAD/FreeCAD/issues/12479

@AjinkyaDahale could you please give it a test and see if there is any problem?

### CODE REVIEW

**Comment by AjinkyaDahale** [comment]
Just a heads up: I'm having build issues right now (likely related to Manjaro update). I'll get back to you once it's all resolved.

**Comment by maxwxyz** [comment]
Nice!
@PaddleStroke do you think Polyline can get in for 1.0?

File: src/Mod/Sketcher/Gui/DrawSketchHandlerPolygon.h
Comment (AjinkyaDahale) [inline src/Mod/Sketcher/Gui/DrawSketchHandlerPolygon.h:1]: Is this related to this PR? If not you could make this a separate PR, or at least a separate commit.

**Review by AjinkyaDahale — COMMENTED** [review]
I'm still struggling to get this working on my device but here's some comment based just on looking at the code changes.

File: src/Mod/Sketcher/Gui/DrawSketchHandlerPolygon.h
Comment (PaddleStroke) [reply src/Mod/Sketcher/Gui/DrawSketchHandlerPolygon.h:1]: This is just a typo fix. It is not related to current PR indeed it would be better as a separate commit.

File: src/Mod/Sketcher/Gui/DrawSketchHandlerBSplineByInterpolation.h
[the remark this answers is not in the document]
Comment (PaddleStroke) [reply src/Mod/Sketcher/Gui/DrawSketchHandlerBSplineByInterpolation.h:1]: First merging DSH was actively done to promote code reuse. With Abdullah we merged the arc DSH and the 3 point arc for this reason. Splitting bspline in 2 DSH was producing a lot of duplication.

Secondly this is the way all the DSH are designed. it is not technically feasible with current framework (that Abdullah designed) to split the construction modes of a tool into separate classes.

[the remark this answers is not in the document]
Comment (AjinkyaDahale) [reply src/Mod/Sketcher/Gui/DrawSketchHandlerBSplineByInterpolation.h:1]: > First merging DSH was actively done to promote code reuse. With Abdullah we merged the arc DSH and the 3 point arc for this reason. Splitting bspline in 2 DSH was producing a lot of duplication.

There's still opportunity to do code reuse while keeping these classes separate.
 
> Secondly this is the way all the new DSH are designed. it is not technically feasible with current framework (that Abdullah designed) to split the construction modes of a tool into separate classes.

This part I can somewhat agree with: we can't have all classes for each mode active at the same time. However, I notice that you have combined modes by periodic vs non-periodic. I would suggest to keep the modes as by poles vs by knots: that way, the points currently chosen make more sense.

File: src/Mod/Sketcher/Gui/DrawSketchDefaultHandler.h [hunk 1]
```diff
@@ -428,20 +428,18 @@ class DrawSketchDefaultHandler: public DrawSketchHandler,
...
         }
     }
 
     void pressRightButton(Base::Vector2d onSketchPos) override
     {
         Q_UNUSED(onSketchPos);
+        rightButtonOrEsc();
+    }
 
+    virtual void rightButtonOrEsc()
```
[the remark this answers is not in the document]
Comment (AjinkyaDahale) [reply src/Mod/Sketcher/Gui/DrawSketchDefaultHandler.h:441]: Maybe `exitFromState`? Agreed that it'll probably not bother us anytime soon. If some left-handed users want to comment they are welcome.

File: src/Mod/Sketcher/Gui/DrawSketchHandlerBSplineByInterpolation.h
[the remark this answers is not in the document]
Comment (PaddleStroke) [reply src/Mod/Sketcher/Gui/DrawSketchHandlerBSplineByInterpolation.h:1]: Periodic is a checkbox. You can toggle it on and off it works well.

**Comment by PaddleStroke** [comment]
1. I do not see the point of changing current situation. It feels that its only to make it possible to separate the code into 2 DSH files. Code should not dictate UI. This is a UI decision, I'd say it's for the @FreeCAD/design-working-group to weight in.
It is true that changing the construction mode reset the tool, as it does with all the DSH. However indeed we could not reset it in this case. It would require a modification of the framework, but it would be needed for polyline anyway.

2. Yes this is not normal. I need to fix this.

**Comment by AjinkyaDahale** [comment]
> I do not see the point of changing current situation. It feels that its only to make it possible to separate the code into 2 DSH files. Code should not dictate UI. This is a UI decision, I'd say it's for the @FreeCAD/design-working-group to weight in.

While I am interested in keeping the classes separate for the refactoring, my rationale for the M behaviour was that it keeps present points the same. If we can also easily toggle between periodic and nonperiodic, particularly through some other shortcut, that works. I'd imagine the mode switches for arc cannot reasonably preserve points as well.

> It would require a modification of the framework, but it would be needed for polyline anyway.

Could you elaborate on the polyline case?

**Comment by PaddleStroke** [comment]
I have fixed the 2 points you mentioned : 
1 - Changing the construction mode does not reset the tool. You can toggle between controlPoint and Knot without loosing your bspline
2 - If you enter a value for the angle or the length, it is then reseted correctly.

**Comment by PaddleStroke** [comment]
> Could you elaborate on the polyline case?

The construction methodes of the polyline case will be Arc / line for the last segment. ie it's not a construction methode for the whole thing. So reseting will not be correct.

**Comment by FEA-eng** [comment]
I wanted to test this but right after entering the second coordinate of the first point, I get a bunch of errors and in some cases FreeCAD even freezes:

```
15:45:47  Unhandled unknown exception caught in GUIApplication::notify.
15:45:48  Unhandled unknown C++ exception in ViewProvider::eventCallback (Event type: SoLocation2Event, object type: SketcherGui::ViewProviderSketch)
```

About FreeCAD info:

```
OS: Ubuntu 22.04.4 LTS (ubuntu:GNOME/ubuntu)
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.37545 (Git)
Build type: Unknown
Branch: sk_bspline_ovp
Hash: 3fae83bfe9425e4f78fc70cbe122177e3a2e55fd
Python 3.10.12, Qt 5.15.3, Coin 4.0.0, Vtk , OCC 7.5.1
Locale: English/United States (en_US)
Installed mods: 
  * OpenDark 2024.3.13
```

**Comment by PaddleStroke** [comment]
You mean using the positioning OVP for the first point right?

**Comment by FEA-eng** [comment]
> You mean using the positioning OVP for the first point right?

Yes, I can enter the first coordinate but applying the second one results in those errors.

**Comment by sliptonic** [comment]
We have assigned this to the 1.0 milestone and moved it back to draft pending resolution of discussion.  Please move out of draft when ready to review again.

**Comment by PaddleStroke** [comment]
The bug has been fixed.

**Comment by FEA-eng** [comment]
> The bug has been fixed.

Now it works. A mode to specify the coordinates of subsequent points (instead of the length and angle of a line connecting them) in this case but this should suffice for now.

File: src/Mod/Sketcher/Gui/DrawSketchHandlerBSpline.h [hunk 2]
```diff
@@ -23,618 +23,1151 @@
...
+
+                // Create subsets of points between C0 knots.
+                // The first point
+                streams.emplace_back();
+                streams.back() << "App.Vector(" << points.front().x << "," << points.front().y
+                               << "),";
+                // Middle points
+                for (size_t i = 1; i < points.size() - 1; ++i) {
+                    streams.back() << "App.Vector(" << points[i].x << "," << points[i].y << "),";
+                    if (multiplicities[i] >= myDegree) {
```
Comment (chennes) [inline src/Mod/Sketcher/Gui/DrawSketchHandlerBSpline.h:201]: Can you please resolve these compiler warnings by casting one or the other of these to an appropriate type? (Applies everywhere you do this comparison)

Comment (PaddleStroke) [reply src/Mod/Sketcher/Gui/DrawSketchHandlerBSpline.h:201]: Sure I'm on it.

File: src/Mod/Sketcher/Gui/DrawSketchHandlerBSpline.h [hunk 3]
```diff
@@ -23,618 +23,1151 @@
...
+    handler->undoLastPoint();
+}
+
+template<>
+void DSHBSplineController::configureToolWidget()
+{
+    if (!init) {  // Code to be executed only upon initialisation
+        toolWidget->setNoticeVisible(true);
+        toolWidget->setNoticeText(
+            QApplication::translate("TaskSketcherTool_c1_bspline", "Press F to undo last point."));
```
Comment (chennes) [inline src/Mod/Sketcher/Gui/DrawSketchHandlerBSpline.h:835]: All the linter noise here might be due to a missing `#include <QApplication>`

**Review by chennes — CHANGES_REQUESTED** [review]
The linter warning is non-critical, but the compiler warnings should be addressed.

**Comment by PaddleStroke** [comment]
Both fixed @chennes
